flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/common/CommonPythonCalc.scala

```diff
@@ -0,0 +1,184 @@
+    val inputNodes = new mutable.LinkedHashMap[RexNode, Integer]()
+    val pythonFunctionInfos = rexCalls.map(createPythonScalarFunctionInfo(_, inputNodes))
+
+    val udfInputOffsets = inputNodes.toArray
+      .map(_._1)
+      .filter(_.isInstanceOf[RexInputRef])
+      .map(_.asInstanceOf[RexInputRef].getIndex)
+    (udfInputOffsets, pythonFunctionInfos)
+  }
+
+  private[flink] def createPythonScalarFunctionInfo(
+      rexCall: RexCall,
+      inputNodes: mutable.Map[RexNode, Integer]): PythonFunctionInfo = rexCall.getOperator match {
+    case sfc: ScalarSqlFunction if sfc.scalarFunction.getLanguage == FunctionLanguage.PYTHON =>
+      val inputs = new mutable.ArrayBuffer[AnyRef]()
+      rexCall.getOperands.foreach {
+        case pythonRexCall: RexCall if pythonRexCall.getOperator.asInstanceOf[ScalarSqlFunction]
+          .scalarFunction.getLanguage == FunctionLanguage.PYTHON =>
+          // Continuous Python UDFs can be chained together
+          val argPythonInfo = createPythonScalarFunctionInfo(pythonRexCall, inputNodes)
+          inputs.append(argPythonInfo)
+
+        case literal: RexLiteral =>
+          inputs.append(
+            convertLiteralToPython.invoke(null, literal, literal.getType.getSqlTypeName))
+
+        case argNode: RexNode =>
+          // For input arguments of RexInputRef, it's replaced with an offset into the input row
+          inputNodes.get(argNode) match {
+            case Some(existing) => inputs.append(existing)
+            case None =>
+              val inputOffset = Integer.valueOf(inputNodes.size)
+              inputs.append(inputOffset)
+              inputNodes.put(argNode, inputOffset)
+          }
+      }
+
+      // Extracts the necessary information for Python function execution, such as
+      // the serialized Python function, the Python env, etc
+      val pythonFunction = new SimplePythonFunction(
+        sfc.scalarFunction.asInstanceOf[PythonFunction].getSerializedPythonFunction,
+        sfc.scalarFunction.asInstanceOf[PythonFunction].getPythonEnv)
+      new PythonFunctionInfo(pythonFunction, inputs.toArray)
+  }
+
+  private[flink] def getPythonScalarFunctionOperator(
+      inputRowTypeInfo: BaseRowTypeInfo,
+      outputRowTypeInfo: BaseRowTypeInfo,
+      udfInputOffsets: Array[Int],
+      pythonFunctionInfos: Array[PythonFunctionInfo],
+      forwardedFields: Array[Int])= {
+    val clazz = Class.forName(PYTHON_SCALAR_FUNCTION_OPERATOR_NAME)
+    val ctor = clazz.getConstructor(
+      classOf[Array[PythonFunctionInfo]],
+      classOf[RowType],
+      classOf[RowType],
+      classOf[Array[Int]],
+      classOf[Array[Int]])
+    ctor.newInstance(
+      pythonFunctionInfos,
+      inputRowTypeInfo.toRowType,
+      outputRowTypeInfo.toRowType,
+      udfInputOffsets,
+      forwardedFields)
+      .asInstanceOf[OneInputStreamOperator[BaseRow, BaseRow]]
+  }
+
+  private [flink] def generatePythonOneInputStream(
+    inputTransform: Transformation[BaseRow],
```

Comment:
Keep four spaces indentation for these parameters. Same for method `createProjectionRexProgram`